File watch overhaul

eatkins opened the pull request:
This is a large pull request that is the culmination of a number of month's work trying to improve the overall io and watch performance of sbt. For the most part, I've broken the commits up so that each one more or less stands on its own. I am doing it as one large pull request just because the overhead of making 20 partially dependent commits feels to high.

This commit uses apis that I developed in the swoval file-tree-views library that improve the performance of traversing the file system tree and also provide a dynamically updating in-memory cache of the file system.

As I describe in the file-tree-views readme, it is surprisingly expensive to descend the file system tree because all of the built in jvm apis require a stat of each file to determine if its a directory or a regular file. In the swoval library, I provide jni native libraries that take advantage of extra information in the system call for listing directories to traverse the tree without stat-ing any of the files. This can reduce the run time of traversal of a directory by 3-5x. It was straightforward to update the PathFinder implementations to use the swoval implementations. Because there is some risk that there is a bad bug in the library, I still allow users to fallback to the built-in java implementations via a system property (I have never seen any unexpected behavior, but I can't rule it out in every environment).

Secondly, I add support to cache a portion of the file system. The motivation is so that we can reduce the latency between the beginning of the compile task to when we actually start compilation. Zinc needs to compare file hashes so right now, sbt has to list all the files and compute the hash. The hash is fast, but the entire process is highly io bound. I seem to get roughly 10K iops on my mac with an ssd so it is perhaps unsurprising that before starting compilation I see sbt doing about 700ms of io before compilation starts in a project with 5000 files. By instead using an in-memory cache of the file system that precomputes the hashes, I've seen that 700ms drop drop to about 100-150. As a bonus, since we are already monitoring the file system to maintain the cache, we can re-use the cache in continuous builds by adding a listener to the cache's observable.

Finally, I make some improvements to the watch api for sbt. I refactor the EventMonitor so that it returns more information to sbt. I use this to make continuous builds testable by allowing the user to provide arbitrary termination conditions which do not depend on reading from System.in. In addition, I am able to implement automatic project reloading when the meta build files change. This is a feature that mill advertises as an improvement over sbt: http://www.lihaoyi.com/post/MillBetterScalaBuilds.html.

I have worked as hard as I could to try and preserve all existing behavior for users. In a number of places, I provide multiple implementations that are user configurable via a system property or sbt setting. In an ideal world, the majority of sbt users notice nothing after the change except that sbt feels snappier than before.

There also was an issue that watching a nested symlinked directory did not work on osx. This was because osx file watchers are all recursive, so I'd added logic to the swoval file events api that pruned any watches for any watches of a descendant of a watched directory. When there are no symlinks, this works great and lowers the resource cost of the watch, but in the presence of a symlinked directory descendant, we'd lose the updates. I fixed this upstream in swoval and I even enhanced it so that updates of symlinked files are detected by the watch service. This enhancement is only available using the swoval file repository and will not be available using the WatchServiceBackedObservable implementation of the FileEventMonitor.

eatkins committed:
Refactor EventMonitor to use a FileEventMonitor

In previous commits, I add the Observable trait along with the
FileEventMonitor trait. These provide the fundamental building blocks
that can be used to implement the simpler EventMonitor trait. To
increase the generality of the provided methods, I add the
WatchServiceBackedObservable class. This implements the Observable
interface mostly using the implementation provided in the
EventMonitorImpl class. Because the implementation logic is largely the
same and the EventMonitor interface is easily expressed using the
FileEventMonitor interface, I was able to replace the EventMonitor apply
method implementation.

As a stand-alone commit, there is not much point to this change. The
WatchServiceBackedObservable is, however, a more useful class than the
EventMonitorImpl and will find use in subsequent commits that replace
the EventMonitor entirely.

One important implementation change though is that there is no longer a
separate thread for polling System.in. The EventMonitor
terminationCondition is now applied by polling the FileEventMonitor for
small periods (i.e. 10.milliseconds) and checking System.in between
polls. This not only reduces the bookkeeping associated with managing
that thread, but it also fixes a bug where in sbt the '~console' command
does not work because the user input thread will detect any input during
the console session (which is sure to include an <enter>). There is
still a background thread for polling the underlying WatchService for
events and invoking the observer callbacks when events are detected.

eatkins committed:
Add FileRepository to io

Throughout the sbt code base, it is necessary to list the contents of a
directory. For large projects, this can become slow because it requires
a lot of file I/O and also may perform computationally expensive work
like computing the sha1 sum of the file contents. It can then be a big
performance win to introduce an in-memory cache of the file system that
is updated as the underlying file contents change.

In this commit, I add three traits that will be used to optimize file
system access:

1) TypedPath -- represents a path on the file system, but may cache some
   of the attributes of the file, e.g. whether or not the file is a
   directory, so that filters may be applied without necessarily having
   to query the file system. Note that the equivalent apis in
   java.io.File or java.nio.file.Files must stat the file which is an
   expensive io operation.

2) FileTreeView -- this provides a simple api for listing the contents
   of a directory. Unlike, say, java.nio.file.Files.walk, the
   FileTreeView.list method returns TypedPath instances, allowing for
   fast filtering.

3) FileTreeDataView -- augments FileTreeView with a listEntries method
   that returns instances of Entry[T], which augment TypedPath with a
   data value. For example, the Entry.value may be the sha-1 sum of a
   source file or it could be the last modified time. Regardless, the
   idea is that it allows the implementation to optionally cache this
   value so that it does not need to do any file system operations to
   return the value to the user. I added this specifically so that I
   could add a zinc external hook in sbt that allows zinc to check which
   source files have changed without having to re-compute these values,
   which is generally bound by disk io, not cpu.

4) FileRepository -- augments FileTreeDataView with the option of
   registering paths for monitoring and adding observers that are
   notified whenever the monitored paths are updated. The repository
   specification does not require caching of the Entry values, though it
   is generally implied.

These traits mirror the
com.swoval.files.{ FileTreeView, FileTreeDataView, FileTreeRepository, TypedPath}
interfaces defined in the swoval file-tree-views library. I do not want
to directly expose the swoval apis to sbt users hence I define the
shadow api. Subsequent commits will provide implementations of these
interfaces.

eatkins committed:
Use Files.isHidden instead of File.isHidden

On linux/osx, a file is considered hidden if its name starts with ".".
Unfortunately, java.io.File.isHidden checks the file system for a hidden
file attribute instead of just checking if the path starts with '.'.
This check takes O(10us) per file, which can lead to noticeable latency
if there are may files to be checked. The java.nio.file.Files.isHidden
method just checks if the path starts with "." on linux/osx and does
whatever it needs to do on windows.

I don't really understand what the point of the file.getName != "."
check is since "." is clearly a hidden file and in most use cases I can
think of, you'd want to exclude it where you exclude other files. But
since I'm not sure what the point is, I'm leaving it alone.

Bonus: annotate type signature of public methods (to quiet intellij)

eatkins committed:
Simplify legacy implementation

I already added sleep to the user input thread in
4435f5b0550a74a4ffb0a835cb616a76a979fff0 so it isn't necessary to add
additional sleep.

eatkins committed:
Refactor SourceModificationWatchSpec

eatkins committed:
Simplify descendantsExcept

Using `---` causes the file system to be traversed twice. The
performance is better using `--` since it will only walk the file tree
once. I also find this easier to read / reason about.

eatkins committed:
Add swoval implementation for (DescendendOrSelf|Child)PathFinder

The built in java apis that allow recursive file tree traversal suffer
from the need to stat every file to determine whether or not it is a
directory or a file. This is very expensive because stat-ing all of the
files can be much more expensive than retrieving them from the OS.
Fortunately, the widely use operating systems (OSX, Windows, Linux) all
provide apis for listing a path that provide limited information about
the returned files. Critically, they will indicate whether the file is a
directory or regular file (and on Linux and OSX, also whether or not the
file is a symbolic link). The built in java apis discard this
information and must make a stat (or windows equivalent) system call on
each returned path, which is very slow. Since there are many situations
where we do not need the other attributes returned by stat, this is a
very fruitful optimization.  Depending on the directory structure, I
have seen the swoval implementation take anywhere between 20% and 50% of
the time to run that the built in java implementations take to run. For
large projects this can be a difference between compilation taking 1
second to start, vs. 250 milliseconds, which is quite noticeable.

Because I have only tested on Windows (32/64 bit), OSX (64 bit) and
Linux (32/64 bit), I still provide an alternative nio based
implementation that can be toggled with the
"sbt.pathfinder.implementation" system property. This also guards
against the possibility that the swoval implementation does not work
correctly for some esoteric directory layouts. Fundamentally both
implementations use the same native os calls (as described above), so
the risk seems small but worth guarding against. The swoval project
itself also has a nio implementation of com.swoval.files.FileTreeViews
and I run tests on all platforms that the native and nio based
implementations return the same results.

eatkins committed:
Add swoval implementations of traits

Now that we have the trait definitions in FileRepository.scala, we need
implementations. I add two basic implementations for FileTreeView and
FileTreeRepository. A FileTreeDataView can be created using an extension
method that converts a FileTreeView into a FileTreeDataView using a
provided converter function.

The new classes DefaultFileTreeView and FileRepositoryImpl effectively
just delegate to an underlying swoval class and perform necessary
conversions between sbt.io and com.swoval.files data types. I also add
the FileRepositorySpec which tests some of the basic semantics of the
FileRepository. I have a significantly more comprehensive test suite of
the swoval implementation of its FileRepository interface that could
eventually be fully ported over to sbt if desired.

eatkins committed:
Bump swoval version

I need to upgrade the swoval library to provide a base implementation
for FileRepository. In the updated swoval version, I renamed the package
that contains the apple file events related code. More importantly, I
also added a java implementation of MacOSXWatchService that is
significantly more robustly tested than the scala implementation that
presently exists in io. To keep things working, I updated the io
implementation to just delegate to the swoval version.

The swoval file-tree-views package is implemented in pure java, which
is nice for io, which must build against a number of scala packages,
including 2.13.X -- a moving target.

eatkins committed:
Add default TypedPath factory

This allows us to make a TypedPath for any java.nio.file.Path.

eatkins committed:
Add HybridPollingFileRepository

There are some cases where it is not possible to monitor a directory
using the os file system notification hooks. For example, if a directory
is actually an nfs mount, this won't work on Linux (I'm not completely
sure under which conditions file system notifications don't work on the
various operating systems). To work around this, I add the
HybridPollingFileRepository which uses a FileRepositoryImpl to cache and
monitor some paths, but doesn't cache or monitor the rest. When listing
paths, if the path to be listed isn't required to be polled, we still
need to partition the results because it's possible that there are
subdirectories that require polling*.

We can make an event monitor for continuous build by combining the
repository event monitor with a polling watch service for the paths that
are not monitored.

* This overlapping use case is discouraged. If it is necessary to mix,
say, nfs mounted directories with local directories in the repository,
then the nfs mounted directories would ideally be in a separate
directory tree from local directories. For example, consider the
following direectory layout:

root (local):
  a (nfs)
  b (local)

Assuming that shouldPoll returns true for directory (a), it would be
fine to register root non-recursively and then to separately register a
and b. I am discouraging registering root recursively, although I make a
best effort at supporting that use case.

I have a hunch that #4179 can be resolved using the
HybridPollingFileRepository to monitor the project that is outside of
the home directory.

eatkins committed:
Rename getNewFiles addNewFiles

Intellij was complaining that this method was a getter that returned
unit.

eatkins committed:
Join PollingThread on close

I noticed that the polling thread was only implicitly halted by setting
the close variable to false. While this did seem to work ok, I generally
think it's good practice to ensure that we join threads in close.
Otherwise, it's possible to have an unnoticed thread leak. With a 5
second join timeout, it will be noticeable if the thread is not quickly
joinable, but it won't halt the application.

Also, we only want to interrupt/join once, so I switched to an
AtomicBoolean for the closed state to use compareAndSet.

eatkins committed:
Add sbt.io.Logger

Previously I had added the EventMonitor.Logger class, but I now want to
remove EventMonitor so I want to move away from using that class.

eatkins committed:
Add FileEventMonitor

In the initial implementation of EventMonitor, I was working with a
largely asynchronous mental model of the system. The EventMonitor was
effectively a black box synchronization point that allowed the user to
wait for anything to happen (but that gave no indication at all about
what happened). Moreover, I conflated two types of events in the
EventMonitor: user input and file system notifications. This conflation
made it somewhat unwieldy and difficult to work with. In this commit, I
add the FileEventMonitor trait which is solely responsible for managing
file system notifications.

While the api is fairly similar to EventMonitor, it has two key
differences:
1) It may be polled for a finite duration. This allows us to interleave
   other work in between polls. More specifically, in the sbt use case,
   we can remove the user input thread that exists in EventMonitor and
   check System.in between polling periods

2) When polled, it returns a sequence of events for every file that was
   updated since the last time it was polled or it blocks until new
   events are received (which are immediately returned) or it times out
   (returning an empty sequence). This allows the user more control over
   what exactly triggers a build. For example, if a file's last modified
   time is changed, but not the contents, the FileEventMonitor may
   return an Update event. The user can choose to ignore this event if
   the contents have not changed. This was not possible with the
   EventMonitor since it discarded all of the information about what
   event actually occurred.

The implementation uses an Observable to watch for events. This is
particularly convenient for tests because we can test the semantics of
the api without actually needing to modify the file system. As a general
rule, tests that actually modify the file system are harder to reason
about because it can be difficult to synchronize the state of the file
system with the state of the program.

A nice property of the new interface is that it is straightforward to
implement anti-entropy quarantining as a decorator on top of a more
simple FileEventMonitor. I uncovered a few tricky edge cases while
working on this, however. For one, it is very common for programs to
implement 'save' with a move of an internal file onto the target
location. This can manifest as a watch process seeing a delete
immediately followed by a creation. When this happens, depending on the
speed of updates, it is actually possible for the program to begin
handling the update before the new file exists. This can be particular
pernicious when paired with anti-entropy because the creation event may
be dropped leaving the program in an inconsistent state. To fix this, I
quarantine ALL deletions for a small period of time. They are removed
from quarantine either when a creation is detected or the quarantine
period expires. That is the logic behind the quarantinedEvents map in
AntiEntropyFileEventMonitor.

eatkins committed:
Add Observers

This class will allow us to selectively add and remove Observers from a
given Observable. The motivation is that a user of a FileRepository may
only want to receive file notification events when their program is in a
certain state. This class also can function as a bus between an Observer
and an Observable allowing for dynamically updating Observers even for
Observables that do not directly support adding and removing Observables
(but perhaps take an Observable as a constructor argument).

eatkins committed:
Remove references to EventMonitor

Now that I've replaced the functionality in EventMonitor with
FileEventMonitor (which has implementations using FileRepository and
WatchServiceBackedObservable), I am replacing the tests of the
EventMonitor interface with tests of the FileEventMonitor interface. The
capabilities that were previously tested all still exist, so no test
cases are created or removed in this commit.

eatkins committed:
Only close the WatchState once

It was previously possible to close the service twice, which we don't
want to do.

eatkins committed:
Add alternative WatchState.empty

For legacy support in sbt, there are situations in which we need to
construct a WatchState that doesn't have a real WatchService associated
with it (to pass into the triggeredMessage and watchingMessage methods).
This could be done inline, but is more convenient to have within io.

eatkins committed:
Deprecate EventMonitor

I'd like to remove EventMonitor entirely, now that I've replaced its
functionality with FileEventMonitor. I'm actually not clear on whether
or not it is kosher to remove this trait right now. It has always only
had private[sbt] visibility. If, however, an old version of sbt > 1.1.4
is run with a new version of io, the program would crash. I'm not sure
if that's possible though because it seems that sbt requires that it is
run against the version of io that was used to compile it. If that is
true, then this commit could just delete EventMonitor.scala, because it
is only directly referenced in this file.

eatkins committed:
Add debug logging to SourceModificationWatchSpec

It can be difficult to reason about why tests failed. This should
hopefully help.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=1.x&zinc=1.x&io=pull/178/head&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

eatkins committed:
Bump test timeout

I was seeing spurious failures in CI that seem to go away if I bump the
timeout.

eatkins committed:
Rewrite PollingWatchService

The PollingWatchServiceSpec was frequently failing both in CI and on my
mac. I found it fairly difficult to reason about because of the mixture
of mutable and immutable data structures so I rewrote the entire thing
from scratch. I made a few significant design decisions:

1) I consolidated the keys and watched maps into a single
   ConcurrentHashMap, registered. To do this, I embed the accepted event
   types in the PollingWatchKey.

2) I rework things to hold the lock for as little as possible. One of
   the reasons for test failures is that both register and the polling
   thread would need the lock and would block each other. Now, we only
   hold the lock when we update the contents of the lastTimestamps map.

3) The PollingThread has no direct state or methods. Since it's an inner
   class of the PollingWatchService, it needs neither.

4) I batch offer the events. One of the issues that I was running into
   was that sometimes a single event would get dropped. This seemed to
   be due to a race condition where the polling thread would be offering
   multiple events for a key while the main thread would be polling the
   service. In some rare events, it was possible for an event to be
   dropped. Now, I pre-compute all of the events, batch add them to the
   key and then batch add all of the keys.

After making these changes, my spurious failures seem to have
disappeared. I also went back and re-worked my new version so that the
layout was similar to the old one. I removed a number of methods that
were only in private classes which required updating the mima filters
since mima does not seem to work how I'd expect with private classes.

eatkins committed:
Remove EventMonitor.legacy method

Rather than carry around the deprecated method that uses a deprecated
api, I inlined a WatchServiceBackedObservable implementation in
SourceModificationWatch.

eed3si9n commented on io/src/main/scala/sbt/internal/io/DefaultFileTreeView.scala:
```diff
@@ -0,0 +1,47 @@
+
+private[sbt] object DefaultFileTreeView extends FileTreeView {
+  private[this] val fileTreeView =
+    if ("nio" == System.getProperty("sbt.pathfinder.implementation"))
```
`sys.props.get(...)` might be better. Also `implementation` in the name feels a bit awkward. I think we tend to call it something like `sbt.pathfinder` with "implementation" implied.

eatkins replied:
Good point. I implemented the swoval library in java and got in the habit of using System.getProperty, so I'll switch to `sys.props.get` and change the name.

eed3si9n commented on io/src/main/scala/sbt/internal/io/DefaultFileTreeView.scala:
```diff
@@ -0,0 +1,47 @@
+import scala.util.Try
+
+private[sbt] object DefaultFileTreeView extends FileTreeView {
+  private[this] val fileTreeView =
```
Not sure how long it takes to initialize fileTreeView, but it might be better to use lazy val here.

eatkins replied:
I think it takes O(2-3ms) to load the swoval jni library if it hasn't been previously loaded. The actual initialization is near instantaneous otherwise.

eed3si9n reviewed (commented):
overall LGTM, I think.
left some minor comments.

eatkins commented:
@eed3si9n Thanks for looking over the PR. I know that it's a lot.

eed3si9n commented on project/Dependencies.scala:
```diff
@@ -18,5 +18,5 @@ object Dependencies {
   val jna = "net.java.dev.jna" % "jna" % "4.5.0"
   val jnaPlatform = "net.java.dev.jna" % "jna-platform" % "4.5.0"
-  val appleFileEvents = "com.swoval" % "apple-file-events" % "1.3.2"
+  val swovalFiles = "com.swoval" % "file-tree-views" % "2.0.0-RC2"
```
It says RC2. Is it waiting on something?

eatkins replied:
I have been reserving the right to change the public api until this PR is good to go. Moving forward, I'm going to enforce binary compatibility of the library in the 2.x series. As soon as we're ready to merge, I will release a 2.0.0 version and update the relevant commit in this PR.

eed3si9n replied:
On develop branch I've added -Xfatal-warnings, so could you rebase against it? After that, I am good to go with this PR.

eatkins replied:
Sure. I published swoval v2.0.0, rebased against origin/develop and updated Dependencies.scala.

eed3si9n commented:
I should've run Dbuild before merging.

```
[sbt212] [error] /home/jenkinssbt/workspace/sbt-validator/target-0.9.14/project-builds/sbt212-0dcf5fdc6c1887745e8427a0496959076954d686/main/src/main/scala/sbt/TemplateCommandUtil.scala:66:12: reference to Logger is ambiguous;
[sbt212] [error] it is imported twice in the same scope by
[sbt212] [error] import sbt.util._
[sbt212] [error] and import sbt.io._
[sbt212] [error]       log: Logger
[sbt212] [error]            ^
[sbt212] [error] /home/jenkinssbt/workspace/sbt-validator/target-0.9.14/project-builds/sbt212-0dcf5fdc6c1887745e8427a0496959076954d686/main/src/main/scala/sbt/TemplateCommandUtil.scala:106:12: reference to Logger is ambiguous;
[sbt212] [error] it is imported twice in the same scope by
[sbt212] [error] import sbt.util._
[sbt212] [error] and import sbt.io._
[sbt212] [error]       log: Logger
[sbt212] [error]            ^
[sbt212] [error] /home/jenkinssbt/workspace/sbt-validator/target-0.9.14/project-builds/sbt212-0dcf5fdc6c1887745e8427a0496959076954d686/main/src/main/scala/sbt/TemplateCommandUtil.scala:134:12: reference to Logger is ambiguous;
[sbt212] [error] it is imported twice in the same scope by
[sbt212] [error] import sbt.util._
[sbt212] [error] and import sbt.io._
[sbt212] [error]       log: Logger
[sbt212] [error]            ^
[sbt212] [info] No documentation generated with unsuccessful compiler run
[sbt212] [error] three errors found
[sbt212] [error] (mainProj / Compile / compileIncremental) Compilation failed
[sbt212] [error] (mainProj / Compile / doc) Scaladoc generation failed
```
